### PR TITLE
Optionally use lua-jsonc parser for parsing vscode-snippets.

### PR DESCRIPTION
**Pull request by L3MON4D3**
The idea is to get the json-parser from the config (default `vim.fn.json_decode`) and we provide the jsonc-parser from https://github.com/actboy168/json.lua, slightly adapted to work more similar to `vim.fn.json_decode`.
I initially though this jsonc-parser would be much slower since it's written in lua, but I can't notice a big difference between the two.
Still, I don't think jsonc is very common, so we should make it opt-in.

### CODE REVIEW

**Comment by L3MON4D3** [comment]
@leiserfg WDYT?

**Comment by leiserfg** [comment]
I think it's better to keep it for ".code-snippets" only, as those are the ones that support jsonc.

**Comment by L3MON4D3** [comment]
Are you certain? I could only find [this](https://code.visualstudio.com/docs/editor/userdefinedsnippets) article, and it seems to make no distinction between the two. (Also, I don't think we even support `*.code-snippets`)

Regardless, allowing comments seems like a nice feature.
A downside with this implementation is that it's a bit all-or-nothing, there's no granularity for parsing files which don't have comments with the regular `vim.fn.json_decode` (although this is alleviated by the lua-parser not being much slower, so being able to write comments doesn't come with much decrease in performance)

**Comment by leiserfg** [comment]
We could use the extensions to use the correct parser, I think that's a good compromise, so `*.json` uses `vim.json.decode` and  `*.jsonc`  uses `jsonc`.
And then in another PR we add `*.code-snippets`  support which is not exactly the same as regular snippets but quite similar.

**Comment by L3MON4D3** [comment]
> We could use the extensions to match the correct parser, I think that's a good compromise, so `*.json` uses `vim.json.decode` and `*.jsonc` uses `jsonc`.

Yes! That's much better :+1:

**Comment by L3MON4D3** [comment]
> And then in another PR we add `*.code-snippets` support which is not exactly the same as regular snippets but quite similar.

Not sure if we need support for that, it doesn't seem to be used as much

File: lua/luasnip/config.lua [hunk 1]
```diff
@@ -151,6 +151,7 @@ local defaults = {
 	load_ft_func = ft_functions.from_filetype_load,
 	-- globals injected into luasnippet-files.
 	snip_env = util.lazy_table({}, lazy_snip_env),
+	json_decode = vim.fn.json_decode,
```
[the remark this answers is not in the document]
Comment (L3MON4D3) [reply lua/luasnip/config.lua:154]: Right!

**Review by leiserfg — APPROVED** [review]

**Review by leiserfg — APPROVED** [review]

**Comment by leiserfg** [comment]
Hey, @L3MON4D3  coming back to this, what do you think of using the jsonc parser of tree-sitter to strip the comments and then parsing the resulting json with  `vim.json`?

**Comment by L3MON4D3** [comment]
Seems like it would work, but I can't really tell what we would gain from this. More speed?

**Comment by leiserfg** [comment]
That's the idea, not sure if it's worth it tho. I'm just thinking out loud.

**Comment by L3MON4D3** [comment]
Ah alright👍
It might be worth it, treesitter is in C after all. Can it do search+replace on its own, or would we have to resort to lua for that?

**Comment by leiserfg** [comment]
This guy did the same:
https://github.com/esensar/nvim-dev-container/blob/main/lua/devcontainer/config_file/jsonc.lua

**Comment by L3MON4D3** [comment]
Oh, yeah that looks good.. Obviously a few string-operations for each comment, doesn't seem to bad, but we'd have to make sure the jsonc-parser is installed on the users machine

But in the end, I doubt it will end up being faster than the jsonc-lua-parser (it would have to be parsed twice+queries+string-ops), but I'd love to be proven wrong :D

**Pull request merged**